created xml IoC container

kwazart opened the pull request:
created xml IoC container
added CSV parser, configuration for execute jar-file
added test sample for convert csv-line to java-entity

Commit:
created xml IoC container
added CSV parser, configuration for execute jar-file
added test sample for convert csv-line to java-entity

stvort commented on src/main/java/com/polozov/springDemo/MainRunner.java:
```diff
@@ -0,0 +1,22 @@
+public class MainRunner {
+    public static void main(String[] args) {
+        ApplicationContext context = new ClassPathXmlApplicationContext("applicationContext.xml");
+        CsvDao dao = context.getBean("parserUtil", CsvDaoImpl.class);
+        QuestionPrinter questionPrinter = context.getBean("questionPrinter", QuestionPrinter.class);
+
+        List<Question> questions = dao.getQuestions();
+        questionPrinter.printQuestionAndGetAnswer(questions);
```
Получать и использовать классы лучше по интерфейсу

kwazart replied:
исправил

stvort commented on src/main/java/com/polozov/springDemo/MainRunner.java:
```diff
@@ -0,0 +1,22 @@
+        CsvDao dao = context.getBean("parserUtil", CsvDaoImpl.class);
+        QuestionPrinter questionPrinter = context.getBean("questionPrinter", QuestionPrinter.class);
+
+        List<Question> questions = dao.getQuestions();
+        questionPrinter.printQuestionAndGetAnswer(questions);
```
Бизнес-логику из main лучше вынести в отдельный сервис/раннер, оставив только получение бина и вызов тестирования. Зависимости бина лучше настроить в spring-context.xml

kwazart replied:
исправил

stvort commented on src/main/java/com/polozov/springDemo/dao/CsvDao.java:
```diff
@@ -0,0 +1,9 @@
+
+import java.util.List;
+
+public interface CsvDao {
```
Одна их основных задач дао, скрыть за общим интерфейсом источник данных. Все что может выдать конкретный источник лучше из интерфейса убрать. Например, слово "Csv"

kwazart replied:
исправил

stvort commented on src/main/java/com/polozov/springDemo/dao/CsvDaoImpl.java:
```diff
@@ -0,0 +1,54 @@
+public class CsvDaoImpl implements CsvDao {
+
+    private final String fileName;
+    // делаю чтобы потренироваться (так как в утильном классе сделал бы статику)
+    private final QuestionConverterUtil converter;
```
Статику как раз лучше не делать. Да и "util" можно из названия убрать. Статические методы создают иллюзию того, что класс разбит согласно ответственности, при этом протестировать его без статики нельзя, как и подменить реализацию

kwazart replied:
исправил

stvort commented on src/main/java/com/polozov/springDemo/dao/CsvDaoImpl.java:
```diff
@@ -0,0 +1,54 @@
+                    questionAndAnswersList.add(Arrays.asList(nextLine));
+                }
+            } catch (FileNotFoundException e) {
+                System.err.println("File " + fileName + " not found");
```
Общение с пользователем не является ответственностью дао, тут можно залоггировать исключение. Если нужен пользовательский вывод, то лучше завернуть исключения в свои и кинуть далее. Затем ближе к `main` поймать их, интерпретировать и вывести сообщение

kwazart replied:
убрал вывод кроме стек трейса (пока без логов)

stvort commented on src/main/java/com/polozov/springDemo/entity/Question.java:
```diff
@@ -0,0 +1,19 @@
+    private String question;
+    private Set<String> answers;
+    private boolean hasFreeAnswer;
+    private String studentAnswer;
```
Ответ студента на вопрос скорее всего не является частью самого вопроса. Он, как и сам вопрос является частью сущности ответа на вопрос (что-то типа `StudentAswer`)

kwazart replied:
убрал ответ студента до лучших времен

stvort commented on src/main/java/com/polozov/springDemo/view/QuestionPrinter.java:
```diff
@@ -0,0 +1,23 @@
+public class QuestionPrinter {
+    public void printQuestionAndGetAnswer(List<Question> questions) {
+        for (Question q : questions) {
+            System.out.println("Question: " + q.getQuestion());
```
Работу с консолью (ввод/вывод, System.in/out) можно вынести в отдельный сервис. Так будет удобнее тестировать классы, где этот ввод/вывод используется. Причем System.in/out такому сервису можно будет тоже передать через конструктор. Такой сервис ничего не будет знать о предметной области и может быть переиспользован

kwazart replied:
так как ответ студента убрал, работу с консолью убрал

stvort commented on src/main/java/com/polozov/springDemo/view/QuestionPrinter.java:
```diff
@@ -0,0 +1,23 @@
+public class QuestionPrinter {
+    public void printQuestionAndGetAnswer(List<Question> questions) {
+        for (Question q : questions) {
+            System.out.println("Question: " + q.getQuestion());
```
Кстати `toString` лучше для вывода не использовать, а сделать отдельный компонент, который переводит сущности в строки

kwazart replied:
так я обращаюсь к стрингу вопроса, а не к самому вопросу. зачем нужен, в данном случае, компонент дополнительный?

stvort commented on src/test/java/com/polozov/springDemo/util/QuestionConverterUtilTest.java:
```diff
@@ -0,0 +1,45 @@
+
+    private static ApplicationContext context;
+
+    @BeforeAll
+    public static void initContext() {
+        context = new ClassPathXmlApplicationContext("applicationContext.xml");
+    }
```
По заданию было нужно сделать юнит-тест. В такого рода тестах участвует исключительно сам класс + моки/заглушки для его зависимостей. Контекст в юнит-тестах лишний

kwazart replied:
оставил только нужное для юнит теста

Commit:
fixed name for classes
used interfaces instead of classes
created main service runner
removed user input (because removed user answers)
removed error output
fixed unit-test (remove context from test)
